### PR TITLE
Add symbolic beam search

### PR DESCRIPTION
**Pull request by junrushao1994**
## Description ##
Symbolic beam search is made possible after enabling control flow operators `mx.sym.contrib.while_loop` (https://github.com/apache/incubator-mxnet/pull/11566) and `mx.sym.contrib.cond` (https://github.com/apache/incubator-mxnet/pull/11760). In this PR, we create a class `HybridBeamSearchSampler`, which could be hybridized to perform beam search.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] Add class `model.beam_search.HybridBeamSearchSampler`
- [x] Allow method `model.beam_search._expand_to_beam_size` to accept Symbols
- [x] Add `_extract_and_flatten_nested_structure` and `_reconstruct_flattened_structure` to flatten and unflatten structure used in decoders
- [x] I slightly modified the unittest for `BeamSearchSampler` and `HybridBeamSearchSampler` to workaround failing testcases causes by `topk`.

## TODO
- [x] Add unittests for `model.beam_search.HybridBeamSearchSampler`
- [x] Rename `vocab_num` to `vocab_size`
- [x] Review docstring again and again, make sure there is nothing wrong.

## Comments ##
- `HybridBeamSearchSampler` requires two extra arguments, `batch_size` and `vocab_size`, compared with `BeamSearchSampler`

### CODE REVIEW

**Comment by mli** [comment]
Job PR-233/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-233/7/index.html

**Comment by leezu** [comment]
Great! Can you make sure the `HybridizedBeamSearchSampler` is tested in `test_beam_search.py` too?

**Comment by junrushao1994** [comment]
@leezu Hey, thank you for the suggestions! So do you prefer I create a new method `test_hybrid_beam_search` or add test code into the existing `test_beam_search`?

Thanks!

**Comment by leezu** [comment]
Would it make sense to make `BeamSearchSampler` are gluon Block? In that case it may be possible to use the same test by parametrizing it with an argument that specifies the sampler implementation and wether to call `.hybridize()`:

```
@pytest.mark.parametrize('hybridize', [True, False])
@pytest.mark.parametrize('sampler', [BeamSearchSampler, HybridizedBeamSearchSampler])
def test_beam_search(hybridize, sampler):
    [...]
```

**Comment by szhengac** [comment]
Good job. But for translation task, the batch_size can change across the iterations, so we cannot simply treat it as an extra parameter.

**Comment by junrushao1994** [comment]
@szhengac It is mandatory for now because static shape inference is required. This issue could be alleviated once symbolic shape is realized.

**Comment by junrushao1994** [comment]
Question: do you guys prefer the name `vocab_num` or `vocab_size`?

**Comment by szha** [comment]
vocab_size, which is a term already used elsewhere in gluonnlp.

**Comment by leezu** [comment]
@junrushao1994 Yes, you can change it to inherit from Block. Just change the ` def __call__(self, inputs, states):` to ` def forward(self, inputs, states):`

**Comment by junrushao1994** [comment]
@szha Seems that there are already many `vocab_num`s in [beam_search.py](https://github.com/dmlc/gluon-nlp/blob/master/gluonnlp/model/beam_search.py#L195), should I change them to `vocab_size` by the way?

**Comment by junrushao1994** [comment]
Waiting for 2 fixes.

1) https://github.com/apache/incubator-mxnet/pull/11942
2) Fix in subgraph cutting

**Comment by junrushao1994** [comment]
It seems that numeral instability and uncertainty would cause unittest to fail. For example, when there are close (or equal) values, `topk` seems unable to produce a result consistent to numpy. So I slightly modify the unittest to let it pass.

**Comment by junrushao1994** [comment]
Still waiting for bug fixes in cutting subgraph

**Comment by szha** [comment]
@junrushao1994 our test environment depends on a specific nightly version. Check under env/ and see if you need to update the date of the nightly build.

**Comment by junrushao1994** [comment]
@zheng-da finds the bug, and just now we submit the fix here: https://github.com/apache/incubator-mxnet/pull/12078

**Comment by junrushao1994** [comment]
There most recent commit fails CI test for the following reason:
> AttributeError: 'LSTM' object has no attribute 'h2h_weight'

Is that an incompatibility issue? @szha

**Comment by szha** [comment]
@junrushao1994 yeah, I'm working on it

**Comment by mli** [comment]
Job PR-233/27 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-233/27/index.html

**Comment by mli** [comment]
Job PR-233/28 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-233/28/index.html

**Comment by junrushao1994** [comment]
@szha This PR has passed CI test, so could you help review the code, especially docstring (my English is pretty bad) Thank you!

File: tests/unittest/test_beam_search.py [hunk 1]
```diff
@@ -196,13 +198,13 @@ def hybrid_forward(self, F, inputs, states):
             return log_probs, states
 
     class RNNLayerDecoder(Block):
```
Comment (szha) [inline tests/unittest/test_beam_search.py]: This can be a hybrid block now

File: tests/unittest/test_beam_search.py [hunk 2]
```diff
@@ -217,25 +219,38 @@ def forward(self, inputs, states):
...
             decoder.hybridize()
             decoder.initialize()
             if hasattr(decoder, 'state_info'):
                 state_info = decoder.state_info()
             else:
                 state_info = None
-            for beam_size, bos_id, eos_id, alpha, K in [(2, 1, 3, 0, 1.0),  (4, 2, 3, 1.0, 5.0)]:
+            if sampler_cls is HybridBeamSearchSampler and decoder_fn is RNNLayerDecoder:
+                # Hybrid beam search does not work on non-hybridizable object
+                continue
```
Comment (szha) [inline tests/unittest/test_beam_search.py]: no need to skip because RNNLayerDecoder can be a hybrid block.

**Comment by mli** [comment]
Job PR-233/32 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-233/32/index.html

**Comment by junrushao1994** [comment]
Thank you guys all the help and suggestions!

**Pull request merged**